From bb809b87b438fd529d41569305bb069e79aed1ae Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 28 Oct 2025 15:44:41 +0800
Subject: [PATCH 124/128] mtk: mac80211: allow setting antenna with radio_idx

Add the support to set TX/RX antenna of a specified radio. In this
case, the nl80211 command should contain radio_idx, and the TX/RX
antenna should not be shifted.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 net/mac80211/cfg.c        |  4 ++--
 net/mac80211/driver-ops.h |  6 +++---
 net/mac80211/trace.h      | 12 ++++++++----
 net/wireless/nl80211.c    | 32 +++++++++++++++++++++++++++++++-
 net/wireless/rdev-ops.h   |  2 +-
 5 files changed, 45 insertions(+), 11 deletions(-)

diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index 4c88746f..4e1b654e 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -4716,10 +4716,10 @@ static int ieee80211_set_antenna(struct wiphy *wiphy, int radio_idx,
 	struct ieee80211_local *local = wiphy_priv(wiphy);
 	int ret;
 
-	if (local->started)
+	if (local->started && radio_idx == -1)
 		return -EOPNOTSUPP;
 
-	ret = drv_set_antenna(local, tx_ant, rx_ant);
+	ret = drv_set_antenna(local, radio_idx, tx_ant, rx_ant);
 	if (ret)
 		return ret;
 
diff --git a/net/mac80211/driver-ops.h b/net/mac80211/driver-ops.h
index e6457c1d..d3d31a90 100644
--- a/net/mac80211/driver-ops.h
+++ b/net/mac80211/driver-ops.h
@@ -769,15 +769,15 @@ static inline void drv_channel_switch(struct ieee80211_local *local,
 }
 
 
-static inline int drv_set_antenna(struct ieee80211_local *local,
+static inline int drv_set_antenna(struct ieee80211_local *local, int radio_idx,
 				  u32 tx_ant, u32 rx_ant)
 {
 	int ret = -EOPNOTSUPP;
 	might_sleep();
 	lockdep_assert_wiphy(local->hw.wiphy);
 	if (local->ops->set_antenna)
-		ret = local->ops->set_antenna(&local->hw, -1, tx_ant, rx_ant);
-	trace_drv_set_antenna(local, tx_ant, rx_ant, ret);
+		ret = local->ops->set_antenna(&local->hw, radio_idx, tx_ant, rx_ant);
+	trace_drv_set_antenna(local, radio_idx, tx_ant, rx_ant, ret);
 	return ret;
 }
 
diff --git a/net/mac80211/trace.h b/net/mac80211/trace.h
index 92a72920..850dd621 100644
--- a/net/mac80211/trace.h
+++ b/net/mac80211/trace.h
@@ -1339,12 +1339,14 @@ DEFINE_EVENT(chanswitch_evt, drv_channel_switch,
 );
 
 TRACE_EVENT(drv_set_antenna,
-	TP_PROTO(struct ieee80211_local *local, u32 tx_ant, u32 rx_ant, int ret),
+	TP_PROTO(struct ieee80211_local *local, int radio_idx, u32 tx_ant,
+		 u32 rx_ant, int ret),
 
-	TP_ARGS(local, tx_ant, rx_ant, ret),
+	TP_ARGS(local, radio_idx, tx_ant, rx_ant, ret),
 
 	TP_STRUCT__entry(
 		LOCAL_ENTRY
+		__field(int, radio_idx)
 		__field(u32, tx_ant)
 		__field(u32, rx_ant)
 		__field(int, ret)
@@ -1352,14 +1354,16 @@ TRACE_EVENT(drv_set_antenna,
 
 	TP_fast_assign(
 		LOCAL_ASSIGN;
+		__entry->radio_idx = radio_idx;
 		__entry->tx_ant = tx_ant;
 		__entry->rx_ant = rx_ant;
 		__entry->ret = ret;
 	),
 
 	TP_printk(
-		LOCAL_PR_FMT " tx_ant:%d rx_ant:%d ret:%d",
-		LOCAL_PR_ARG, __entry->tx_ant, __entry->rx_ant, __entry->ret
+		LOCAL_PR_FMT " radio_idx:%d tx_ant:%d rx_ant:%d ret:%d",
+		LOCAL_PR_ARG, __entry->radio_idx, __entry->tx_ant,
+		__entry->rx_ant, __entry->ret
 	)
 );
 
diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index c3459343..431dcd95 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -3812,12 +3812,18 @@ static int nl80211_set_wiphy_radio(struct genl_info *info,
 				   struct cfg80211_registered_device *rdev,
 				   int radio_idx)
 {
+	const struct wiphy_radio *radio;
 	u32 rts_threshold = 0, old_rts, changed = 0;
 	int result;
 
 	if (!rdev->ops->set_wiphy_params)
 		return -EOPNOTSUPP;
 
+	if (radio_idx < 0 || radio_idx >= rdev->wiphy.n_radio)
+		return -EINVAL;
+
+	radio = &rdev->wiphy.radio[radio_idx];
+
 	if (info->attrs[NL80211_ATTR_WIPHY_RTS_THRESHOLD]) {
 		rts_threshold = nla_get_u32(
 				info->attrs[NL80211_ATTR_WIPHY_RTS_THRESHOLD]);
@@ -3832,7 +3838,31 @@ static int nl80211_set_wiphy_radio(struct genl_info *info,
 	if (result)
 		rdev->wiphy.radio_cfg[radio_idx].rts_threshold = old_rts;
 
-	return 0;
+	if (info->attrs[NL80211_ATTR_WIPHY_ANTENNA_TX] &&
+	    info->attrs[NL80211_ATTR_WIPHY_ANTENNA_RX]) {
+		u32 tx_ant, rx_ant, radio_antenna_mask;
+
+		if (!radio->antenna_mask || !rdev->ops->set_antenna)
+			return -EOPNOTSUPP;
+
+		/* Restore the shifted radio->antenna_mask */
+		radio_antenna_mask = radio->antenna_mask >> __ffs(radio->antenna_mask);
+		tx_ant = nla_get_u32(info->attrs[NL80211_ATTR_WIPHY_ANTENNA_TX]);
+		rx_ant = nla_get_u32(info->attrs[NL80211_ATTR_WIPHY_ANTENNA_RX]);
+
+		/* reject antenna configurations which don't match the
+		 * available antenna masks, except for the "all" mask */
+		if ((~tx_ant && (tx_ant & ~radio_antenna_mask)) ||
+		    (~rx_ant && (rx_ant & ~radio_antenna_mask)))
+			return -EINVAL;
+
+		tx_ant = tx_ant & radio_antenna_mask;
+		rx_ant = rx_ant & radio_antenna_mask;
+
+		result = rdev_set_antenna(rdev, radio_idx, tx_ant, rx_ant);
+	}
+
+	return result;
 }
 
 static int nl80211_set_wiphy(struct sk_buff *skb, struct genl_info *info)
diff --git a/net/wireless/rdev-ops.h b/net/wireless/rdev-ops.h
index 31f89933..9dc61a1f 100644
--- a/net/wireless/rdev-ops.h
+++ b/net/wireless/rdev-ops.h
@@ -866,7 +866,7 @@ static inline int rdev_set_antenna(struct cfg80211_registered_device *rdev,
 {
 	int ret;
 	trace_rdev_set_antenna(&rdev->wiphy, radio_idx, tx_ant, rx_ant);
-	ret = rdev->ops->set_antenna(&rdev->wiphy, -1, tx_ant, rx_ant);
+	ret = rdev->ops->set_antenna(&rdev->wiphy, radio_idx, tx_ant, rx_ant);
 	trace_rdev_return_int(&rdev->wiphy, ret);
 	return ret;
 }
-- 
2.45.2

